File: java-suite/src/main/java/de/fraunhofer/iem/reflection/TypicalExample.java
package de.fraunhofer.iem.reflection;

import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * Demonstrates the method and field access through Java reflection
 *
 * @author Ranjith Krishnamurty
 */
public class TypicalExample {

    public static int publicField = 20;

    public static void publicMethod() {
        System.out.println("This is a method with the name publicMethod");
    }

    private static void fieldAccess(Class<?> clz) throws NoSuchMethodException, InvocationTargetException, InstantiationException, IllegalAccessException {
        Field[] fields = clz.getDeclaredFields(); // getDeclaredField can be private or public

        for (Field field : fields) {
            System.out.println(field.getName());
            Object ob = clz.getDeclaredConstructor().newInstance();

            System.out.println(field.get(ob));

            field.set(ob, 50);

            System.out.println(field.get(ob));
        }
    }

    private static void methodAccess(Class<?> clz) throws NoSuchMethodException, InvocationTargetException, InstantiationException, IllegalAccessException {
        Method[] methods = clz.getDeclaredMethods(); // getDeclaredMethods can be private or public

        for (Method method : methods) {
            if (method.getName().equals("publicMethod")) {
                System.out.println(method.getName());
                Object ob = clz.getDeclaredConstructor().newInstance();

                method.invoke(ob);
            }
        }
    }

    public static void main(String[] args) throws ClassNotFoundException, NoSuchFieldException, NoSuchMethodException, InvocationTargetException, InstantiationException, IllegalAccessException {
        Class<?> clz = Class.forName("de.fraunhofer.iem.reflection.TypicalExample");

        fieldAccess(clz);
        methodAccess(clz);
    }
}

File: kotlin-suite/src/main/kotlin/de/fraunhofer/iem/reflection/TypicalExample.kt
package de.fraunhofer.iem.reflection

import kotlin.reflect.KClass
import kotlin.reflect.KMutableProperty
import kotlin.reflect.full.createInstance
import kotlin.reflect.full.memberFunctions
import kotlin.reflect.full.memberProperties

/**
 * Demonstrates the method and property access through Kotlin reflection
 *
 * @author Ranjith Krishnamurthy
 */
class TypicalExample {

    var publicField = 20

    fun publicMethod() {
        println("This is a method with the name publicMethod")
    }

    /**
     * Demonstrates the property access through KClass
     */
    fun propertyAccess(clz: KClass<*>) {
        val properties = clz.memberProperties

        for (property in properties) {
            println(property.name)
            val ob = clz.createInstance()
            println(property.getter.call(ob))

            if (property is KMutableProperty<*>) {
                property.setter.call(ob, 50)
            }

            println(property.getter.call(ob))
        }
    }

    /**
     * Demonstrates the property access directly from the class
     */
    fun directPropertyAccess() {
        val ob = TypicalExample::class.createInstance()
        println(TypicalExample::publicField.name)
        println(TypicalExample::publicField.getter.call(ob))

        TypicalExample::publicField.setter.call(ob, 50)
        println(TypicalExample::publicField.getter.call(ob))
    }

    /**
     * Demonstrates the method access through KClass
     */
    fun methodAccess(clz: KClass<*>) {
        val methods = clz.memberFunctions

        for (method in methods) {
            if (method.name == "publicMethod") {
                println(method.name)
                val ob = clz.createInstance()
                method.call(ob)
            }
        }
    }

    /**
     * Demonstrates the method access directly from the class
     */
    fun directMethodAccess() {
        val ob = TypicalExample::class.createInstance()
        println(TypicalExample::publicMethod.name)
        TypicalExample::publicMethod.call(ob)
    }
}

fun main() {
    val typicalExample = TypicalExample()
    println("*** Property Access ***")
    typicalExample.propertyAccess(TypicalExample::class)
    println("\n*** Method Access ***")
    typicalExample.methodAccess(TypicalExample::class)
    println("\n*** Direct Property Access ***")
    typicalExample.directPropertyAccess()
    println("\n*** Direct Method Access ***")
    typicalExample.directMethodAccess()
}